Allow for passwordless certs

messiaen opened the pull request:
We wanted to run some automated scripts using pypki2, so I added the ability to configure passwordless certificates in this way.

I added a "nopass" key to each of the "pem" and "p12" configs.  If "nopass" is "true" then the user will not be prompted for a password.  "nopass" is set to true during initial configuration if the user enters a empty password.

To avoid being prompted for my p12 certs even though I had configured my pem certs, I changed the way the default certificate loader is chosen.
1. The user can specify the "default" key to be "pem" or "p12".  If that key is set, choose the loader associated with its value
2. If the .mypki file only has one of "pem" or "p12" then try that one first.

messiaen committed:
Allow for passwordless certs

	- Add "nopass" key to pem and p12 config. If "nopass": "ture",
	  the user will not be prompted for a password
	- When configuring certs the user can simply hit enter at the
	  password prompt and the cert will be configured a passwordless
	  certs, and the user will not be prompted subsequently

messiaen committed:
Better loader choosing

	- Allow user to set default, with the "default" key in their
	  config. (e.g. {"default": "pem", "pem": {...}, "p12": {...}
	- If only one of "pem" or "p12" is present, try that one first

billallen256 commented:
@kylekyle are you ok with adding a nopass option for each type of cert specified in .mypki?  It's for people that want to use passwordless certs in non-interactive scripts.

For example:
```
{ "p12": { "path": "/some/path.p12", "nopass": "true" }}
```

The default would be false, if nopass is not specified.  Thoughts?

kylekyle commented:
@gershwinlabs: If you don't specify a password, it already assumes you don't need one.

billallen256 commented:
@messiaen there's probably a way to have the code check if the cert at the specified path is encrypted.  If it isn't encrypted, then we can skip the password prompt.  It also saves us from having to keep extra info in the .mypki config file as nopass.

kylekyle commented:
@gershwinlabs: I definitely lied. The config is loader specific, so PEM doesn't need a password, but P12 always does. Yes, adding a nopass option for P12 is a good idea.

billallen256 commented:
@kylekyle: Does the Ruby version do any checking of the P12 or PEM cert to see if it's actually encrypted before bothering the user with a password prompt?

kylekyle commented:
The PEM loader checks the header for the word "ENCRYPTED". I didn't find a good way to check P12s, but that doesn't mean there isn't one. You can check out the loaders folder in the MyPKI project.

messiaen committed:
Remove nopass option from config

	- Check if certificate is encrypted by attempting to decrypt it
	  with an empty password (i.e. call load_privatekey and see if it
	  fails with en empty password)
	- Only prompt the user for a password when necessary

messiaen commented:
A simple way to check if a cert is encrypted is to try to decrypt it with an empty password.

Only if decrypting with an empty password fails, would we prompt the user for a password.

billallen256 commented:
@messiaen Nice work.

@kylekyle No need to add a "nopass" option, though this does add an optional top-level key called "default" that allows a default loader to be set.  The current valid values are "pem" and "p12".  Would this cause issues for mypki?

billallen256 commented:
This feature was resolved and released as version 0.8.0 (8cff4cc5998955ed4cad0926bf68e55c1bca288b).
